runtime/common/src/main/java/edu/snu/coral/runtime/common/optimizer/pass/runtime/DataSkewRuntimePass.java

```diff
@@ -130,15 +156,33 @@ public PhysicalPlan apply(final PhysicalPlan originalPlan, final Map<String, Lis
           currentAccumulatedSize += aggregatedMetricData.get(finishingHashValue);
           finishingHashValue++;
         }
-        // Go back once if we came too far.
-        if (currentAccumulatedSize - idealAccumulatedSize
-            > idealAccumulatedSize - (currentAccumulatedSize - aggregatedMetricData.get(finishingHashValue - 1))) {
-          finishingHashValue--;
-          currentAccumulatedSize -= aggregatedMetricData.get(finishingHashValue);
+
+        long finalSize;
+        if (i == 1) {
+          finalSize = currentAccumulatedSize;
+        } else {
+          long currentSize = currentAccumulatedSize - sizePerTaskGroup.stream().mapToLong(l -> l).sum();
+          long oneStepBack = currentSize - aggregatedMetricData.get(finishingHashValue - 1);
+
+          // If the accumulated size for this TaskGroup exceeds upperBoundSize
+          // and taking off for one hash range doesn't violate lowerBoundSize, go one step back.
```

Comment:
Before, we compared the difference of currentAccumulatedSize and oneStepBack from the idealAccumulatedSize, to pick the one with the less difference. Now, we simply see if taking off one hash range doesn't violate the lowerBoundSize. Is this change better for evening out data?